File: src/main/java/com/sedmelluq/discord/lavaplayer/source/youtube/YoutubeSignatureCipher.java
package com.sedmelluq.discord.lavaplayer.source.youtube;

import java.util.ArrayList;
import java.util.List;

/**
 * Describes one signature cipher
 */
public class YoutubeSignatureCipher {
  private final List<YoutubeCipherOperation> operations = new ArrayList<>();

  /**
   * @param text Text to apply the cipher on
   * @return The result of the cipher on the input text
   */
  public String apply(String text) {
    StringBuilder builder = new StringBuilder(text);

    for (YoutubeCipherOperation operation : operations) {
      switch (operation.type) {
        case SWAP:
          int position = operation.parameter % text.length();
          char temp = builder.charAt(0);
          builder.setCharAt(0, builder.charAt(position));
          builder.setCharAt(position, temp);
          break;
        case REVERSE:
          builder.reverse();
          break;
        case SLICE:
        case SPLICE:
          builder.delete(0, operation.parameter);
          break;
        default:
          throw new IllegalStateException("All branches should be covered");
      }
    }

    return builder.toString();
  }

  /**
   * @param operation The operation to add to this cipher
   */
  public void addOperation(YoutubeCipherOperation operation) {
    operations.add(operation);
  }

  /**
   * @return True if the cipher contains no operations.
   */
  public boolean isEmpty() {
    return operations.isEmpty();
  }
}

File: main/src/main/java/lavaplayer/source/youtube/YoutubeSignatureCipher.kt
package lavaplayer.source.youtube

import lavaplayer.source.youtube.YoutubeCipherOperationType.*

/**
 * Describes one signature cipher
 */
class YoutubeSignatureCipher {
    private val operations = mutableListOf<YoutubeCipherOperation>()

    var scriptTimestamp = ""

    /**
     * @return True if the cipher contains no operations.
     */
    val isEmpty: Boolean
        get() = operations.isEmpty()

    /**
     * @param text Text to apply the cipher on
     * @return The result of the cipher on the input text
     */
    fun apply(text: String): String {
        val builder = StringBuilder(text)
        for (operation in operations) {
            when (operation.type) {
                SWAP -> {

                    val position = operation.parameter % text.length
                    val temp = builder[0]
                    builder.setCharAt(0, builder[position])
                    builder.setCharAt(position, temp)
                }
                REVERSE -> builder.reverse()
                SLICE, SPLICE -> builder.delete(0, operation.parameter)
            }
        }

        return builder.toString()
    }

    /**
     * @param operation The operation to add to this cipher
     */
    fun addOperation(operation: YoutubeCipherOperation) {
        operations.add(operation)
    }
}
